[Work in progress] Add XMC4200 Distance2Go board

yennster opened the pull request:
Need to fix pins_arduino.h

yennster commented:
@mhollfelder @techpaul Do you have any suggestions / ideas on how I can get the XMC4200 Distance2Go board to work within Arduino? 

I would like to simply pull the IQ raw data off of the XMC4200 Distance2Go board as the Arduino example for the XMC1300 Sense2GoL does here: https://www.infineon.com/cms/en/product/promopages/makeradar/makeradar-school/programming-tutorial/#readrawdata

techpaul commented on boards.txt:
```diff
@@ -261,6 +261,50 @@ XMC1300_Sense2GoL.menu.LIB.DSP.library.selected=-DARM_LIB_CMSIS_DSP
+XMC4200_Distance2Go.upload.tool=xmcprog
+XMC4200_Distance2Go.upload.speed=115200
+XMC4200_Distance2Go.upload.resetmethod=ck
+XMC4200_Distance2Go.upload.maximum_size=512000
```
Value is 500 kB and device is 256 kB value should be 262144

techpaul commented on boards.txt:
```diff
@@ -261,6 +261,50 @@ XMC1300_Sense2GoL.menu.LIB.DSP.library.selected=-DARM_LIB_CMSIS_DSP
+XMC4200_Distance2Go.build.f_cpu=80000000L
+XMC4200_Distance2Go.build.board=ARM_XMC
+XMC4200_Distance2Go.build.board.version=4200
+XMC4200_Distance2Go.build.board.type=Q48x256
```
If device is really a 48 pin device there are MANY pins missing in pins_arduino.h

techpaul commented on variants/XMC4200/config/XMC4200_Distance2Go/pins_arduino.h:
```diff
@@ -0,0 +1,226 @@
+#define NUM_ANALOG_INPUTS   6
+#define NUM_PWM             6
+#define NUM_LEDS            3
+#define NUM_INTERRUPT       2
```
You can have as many interrupt pins as the device supports

techpaul commented on variants/XMC4200/config/XMC4200_Distance2Go/pins_arduino.h:
```diff
@@ -0,0 +1,226 @@
+#define NUM_INTERRUPT       2
+#define NUM_SERIAL          1
+#define NUM_TONE_PINS       8
+#define NUM_TASKS_VARIANT   16
```
Considering device size and clock rate I would consider making this 16 or 32, these are software driven tasks from systick counts down to ms resolution

techpaul commented on variants/XMC4200/config/XMC4200_Distance2Go/pins_arduino.h:
```diff
@@ -0,0 +1,226 @@
+#define A3   3
+#define A4   4
+#define A5   5
+
```
Is the implementation capable of more Analogue inputs if so put ALL here and in other arrays

techpaul commented on variants/XMC4200/config/XMC4200_Distance2Go/pins_arduino.h:
```diff
@@ -0,0 +1,226 @@
+XMC_ADC_t mapping_adc[] = {
+	{VADC, 14, VADC_G0, 0, 3, DISABLED},			// CH_I
+	{VADC, 15, VADC_G0, 0, 4, DISABLED}				// CH_Q
+};
```
Missing at least A2 to A5 referenced above

techpaul commented on variants/XMC4200/config/XMC4200_Distance2Go/pins_arduino.h:
```diff
@@ -0,0 +1,226 @@
+/*
```
This file appears to have started from Sense2GoL file which has many errors in it I am working on an improved one for that

techpaul commented on variants/XMC4200/config/XMC4200_Distance2Go/pins_arduino.h:
```diff
@@ -0,0 +1,226 @@
+	/* 0  */ 	//{CCU40, CCU40_CC40, 0, 0, CCU40_IN0_U0C0_DX2INS}, 
+	/* 1  */ 	//{CCU40, CCU40_CC40, 0, 1, CCU40_IN0_P0_0} 
+    };
+
```
CCU4 and CCU8 missing supposedly 6 PWM and no PWM pin definitions and mappings

techpaul commented on variants/XMC4200/config/XMC4200_Distance2Go/pins_arduino.h:
```diff
@@ -0,0 +1,226 @@
+#define digitalPinToInterrupt(p)    ((p) == 2 ? 0 : ((p) == 32 ? 1 : NOT_AN_INTERRUPT))
+
+#ifdef ARDUINO_MAIN
+
```
Missing PWM4 and PWM8 to port and pin mapping

techpaul commented on variants/XMC4200/linker_script.ld:
```diff
@@ -0,0 +1,281 @@
+ * Linker file for the GNU C Compiler v1.8
+ * Supported devices: XMC4200-F64x256
+ *                    XMC4200-Q48x256
+ *
```
Supported devices is ANY 4200 with 256kB Flash and 40kB RAM, Package is IRRELEVANT to linker

techpaul commented:
Get boards.txt and fully define all pins that can be used in any way defined in pins_arduino.h

If serial working go from there

Many simple tests and examples exist to check operation like tonetest, dietempmeasurement and in tests area testdefines to check you are creating correct details.

techpaul commented:
> @mhollfelder @techpaul Do you have any suggestions / ideas on how I can get the XMC4200 Distance2Go board to work within Arduino?
> 
> I would like to simply pull the IQ raw data off of the XMC4200 Distance2Go board as the Arduino example for the XMC1300 Sense2GoL does here: https://www.infineon.com/cms/en/product/promopages/makeradar/makeradar-school/programming-tutorial/#readrawdata

I am personally AMAZED that works as the line about turning BGT_ON to on does NOTHING to the RADAR chip as it refers to a PWM output  on port0.6 (27) which is labelled WRONGLY in pins_arduino.h.

I would expect for the Sense2GoL the signal TX_ON on Port0.0 (3) to actually do something and it does not get referenced anywhere.

yennster commented:
> > @mhollfelder @techpaul Do you have any suggestions / ideas on how I can get the XMC4200 Distance2Go board to work within Arduino?
> > I would like to simply pull the IQ raw data off of the XMC4200 Distance2Go board as the Arduino example for the XMC1300 Sense2GoL does here: https://www.infineon.com/cms/en/product/promopages/makeradar/makeradar-school/programming-tutorial/#readrawdata
> 
> I am personally AMAZED that works as the line about turning BGT_ON to on does NOTHING to the RADAR chip as it refers to a PWM output on port0.6 (27) which is labelled WRONGLY in pins_arduino.h.
> 
> I would expect for the Sense2GoL the signal TX_ON on Port0.0 (3) to actually do something and it does not get referenced anywhere.

@techpaul Yeah I was wondering how `BGT_ON` works. Do you know where I can find a more succinct  list of all pins / mappings for the Distance2Go board? I downloaded the schematics file from the Infineon toolbox but it's a bit of a headache to match all of the pins to their corresponding usage (EXT INT, GPIO, etc.)

techpaul commented:
For definitions and what capabilities I use the board manuls,, then the datasheet and reference manual for the micro. Also look at how other similar devices are done in their files for similar functions.

I would check for similarities to XMC4400_Platform2GO as that has support for similar micro, check on differences.

yennster commented:
@techpaul Made a bunch of changes, please check them out. I am trying to compile now with the arduino CLI but I still get the following error: 

```
In file included from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\Arduino.h:263:0,
                 from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\HardwareSerial.h:36,
                 from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\HardwareSerial.cpp:22:
C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0/variants/XMC4200/config/XMC4200_Distance2Go/pins_arduino.h:251:8: error: 'HardwareSerial' does not name a type
 extern HardwareSerial Serial;
        ^
In file included from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\Arduino.h:263:0,
                 from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\Main.cpp:27:
C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0/variants/XMC4200/config/XMC4200_Distance2Go/pins_arduino.h:228:41: error: conflicting declaration of 'void serialEvent()' with 'C' linkage
 void serialEvent( ) __attribute__((weak));
                                         ^
In file included from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\Arduino.h:255:0,
                 from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\Main.cpp:27:
C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\HardwareSerial.h:30:6: note: previous declaration with 'C++' linkage
 void serialEvent() __attribute__((weak));
      ^
```

It's not including `HardwareSerial` for this new variant or something?

techpaul commented:
It is late veening here as soon as I can, I will look this over in more detail, my initial checks using Notepad++ and collapsing #if and #endif pairing is casuing for some reason the  #ifdef ARDUINO_MAIN collapsing to collapse the REST of the file as IF missing a #endif.

Suggest some comment or other missing ';' error somewhere appears to be around port pin mapping but not found it yet

yennster commented:
@techpaul I cleaned up pins_arduino.h and fixed the comments, etc. and I'm still seeing the errors:

```
PS C:\Local-Windows\ei-sketch-1dconv> arduino-cli compile --fqbn Infineon:arm:XMC4200_Distance2Go ei-sketch-1dconv -e --log-file .\log.txt --log-format text --log-level trace --output-dir .
In file included from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\Arduino.h:261:0,
                 from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\HardwareSerial.h:36,
                 from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\HardwareSerial.cpp:22:
C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0/variants/XMC4200/config/XMC4200_Distance2Go/pins_arduino.h:211:8: error: 'HardwareSerial' does not name a type
 extern HardwareSerial Serial;
        ^
In file included from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\Arduino.h:261:0,
                 from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\Main.cpp:27:
C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0/variants/XMC4200/config/XMC4200_Distance2Go/pins_arduino.h:188:41: error: conflicting declaration of 'void serialEvent()' with 'C' linkage
 void serialEvent( ) __attribute__((weak));
                                         ^
In file included from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\Arduino.h:256:0,
                 from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\Main.cpp:27:
C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\HardwareSerial.h:30:6: note: previous declaration with 'C++' linkage
 void serialEvent() __attribute__((weak));
      ^

Error during build: exit status 1
```

Any other ideas? I'm scratching my head pretty hard

techpaul commented:
Ignore the second two errors as they are because pins_arduino.h is not being included correctly as is shown by the first error. The problem looks like #if and #endif pairing.

Unfortunately the include files are a bit of a tree of includes inculding includes ad infinitum... So from
~~~
In file included from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\Arduino.h:261:0,
                 from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\HardwareSerial.h:36,
                 from C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0\cores\HardwareSerial.cpp:22:
C:\Users\jenny\AppData\Local\Arduino15\packages\Infineon\hardware\arm\1.4.0/variants/XMC4200/config/XMC4200_Distance2Go/pins_arduino.h:211:8: error: 'HardwareSerial' does not name a type
 extern HardwareSerial Serial;
~~~
HardwareSerial.cpp --> HardwareSerial.h --> Arduino.h --> pins_arduino.h [--> 4200.h --> system_4200.h]

So files you have changed are 4200.h system_4200.h and pins_arduino.h I would start there if I get time I will look at 4200.h

Try inserting same #warning messages to see which parts are included

For something like this I would normally start in Arduino IDE with a blank sketch and compile/verify that as first test

yennster commented:
@techpaul Think I figured it out, the version of XMC-for-Arduino that my Arduino15 folder was using was version 1.4.0, something changed between 1.4.0 and 1.5.0, and 1.5.0 is what I was creating the pins_arduino.h file for and that's why I was seeing that error. Now I am building successfully: 

```
PS C:\Local-Windows\ei-sketch-1dconv> arduino-cli compile --fqbn Infineon:arm:XMC4200_Distance2Go ei-sketch-1dconv -e --log-file .\log.txt --log-format text --log-level trace --output-dir .
Sketch uses 42852 bytes (16%) of program storage space. Maximum is 262144 bytes.
Global variables use 6392 bytes of dynamic memory.
```

I knew it was a simple issue!!!! now to see if the binary actually runs on the board...

techpaul commented:
Glad you got to bottom of it Hope we were at least some help

yennster commented:
@techpaul A simple serial print program isn't working now... Sigh... Google searching doesn't help as many people haven't been able to get the UART to work on the XMC4200 either:  https://www.infineonforums.com/threads/6156-Xmc-4200-uart

techpaul commented:
First thing to check is pins_arduino.h for mainly XMC_UART_t XMC_UART_0 
* pin routing in the  .rx/.tx_config
* input source routing .input_source_dx<x>
* interrupts all way back from irq_num and irq_service_request  **All** the way back to vector tables

The usual problems are pin mapping and interrupts even for serial.print

techpaul commented:
A few simple Changes I would suggest pins_arduino.h
~~~
#define LED_BUILTIN 25  //Standard Arduino LED: Used LED1
#define LED1        25  //Additional LED1 P2.0
#define GND         14  //GND
~~~
To 
~~~
#define LED_BUILTIN 15  //Standard Arduino LED: Used LED1
#define LED1        15  //Additional LED1 P2.0
#define GND        29  //GND
~~~
You have defined Arduino pin 15 (P2.0) as the LED

GND iis used as one more than valid  Arduino pins often used as limits checks and error ignoring in pin assignment,
(probably could be made into a sizeof operation)

Unfortunately not found circuit reference for the Distance2Go board to check what is happening and what GPIO are still available.

yennster commented:
@techpaul I am trying every single combination of UART pins within DAVE with the UART App for the XMC4200-256 board in the DAVE IDE, and nothing ever gets printed to Serial with the below program:

```
/*
 * main.c
 *
 *  Created on: 2021 Jan 18 16:57:33
 *  Author: jenny
 */

#include <DAVE.h>
#include <xmc_common.h>

//extern void initialise_monitor_handles(void);

const char message[] = "Hello world!!!\n";

/* This "wires" the putchar/printf functions to transmit to UART_0 */
/*int _write(int file, uint8_t *buf, int nbytes)
{
	if(UART_Transmit(&UART_0, buf, nbytes) == UART_STATUS_SUCCESS) {
		while(UART_0.runtime->tx_busy) {
		}
	}
	return nbytes;
}*/

/* This "wires" the getchar function to receive from UART_0 */
/*int _read(int file, uint8_t *buf, int nbytes)
{
	if(UART_Receive(&UART_0, buf, 1) != UART_STATUS_SUCCESS) {
		nbytes = 0;
	} else {
		nbytes = 1;
	}
	return nbytes;
}*/

int main(void)
{
	DAVE_STATUS_t status;

	//initialise_monitor_handles();

	status = DAVE_Init();           /* Initialization of DAVE APPs  */

	if(status != DAVE_STATUS_SUCCESS)
	{
		/* Placeholder for error handler code. The while loop below can be replaced with an user error handler. */
		XMC_DEBUG("DAVE APPs initialization failed\n");

		while(1U)
		{

		}
	}

	while(1U)
	{
		int x = 0;
		while (1U) {
			char buffer[20];
			sprintf(buffer, "%d \n", x);
			x++;
			UART_Transmit(&UART_0, buffer, strlen(buffer));
		}
	}
}
```

If I switch the board to XMC1302 Sense2GoL in the DAVE IDE with the same program as above and a UART Dave App it runs and prints to Serial just fine.

I'm beginning to suspect it's an Infineon issue? Or maybe my board is busted?

yennster commented:
I've created an Infineon forum post here: https://www.infineonforums.com/threads/12254-XMC4200-Distance2Go-UART-not-working

yennster commented:
FYI @techpaul leaving this PR to die for now as with help from @mkiessling-ifx I was able to get an example working with Edge Impulse and the Sense2GoL-Pulse board on the 24Ghz-radar branch

techpaul commented:
Well hopefully someone cn pick up from wht you have achieved so far.

Hope your project goes well
